### PR TITLE
Fixed TextBox OutOfRangeException

### LINKED ISSUE #1954: TextBox throws System.ArgumentOutOfRangeException after text input
Occasional texbox exceptions ArgumentOutOfRangeException after text input.
Hard to repro with user actions only, Text must be bound to regular changing value.

StackTrace:
 at System.String.Substring(Int32 startIndex, Int32 length)
   at Avalonia.Controls.TextBox.DeleteSelection()
   at Avalonia.Controls.TextBox.HandleTextInput(String input)
   at Avalonia.Controls.TextBox.OnTextInput(TextInputEventArgs e)
   at Avalonia.Interactivity.RoutedEvent`1.<>c__DisplayClass1_0`1.<AddClassHandler>b__0(Object sender, RoutedEventArgs e)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Avalonia.Interactivity.RoutedEvent.<>c__DisplayClass25_0.<AddClassHandler>b__0(Tuple`2 args)
   at System.Reactive.ObserverBase`1.OnNext(T value)

### PR DESCRIPTION
**Pull request by donandren**
- What does the pull request do?
Fixes occasional textbox exceptions issue #1954
- What is the current behavior?
Throws exceptions
- What is the updated/expected behavior with this PR?
Textbox have valid state and don't throw exceptions
- How was the solution implemented (if it's not obvious)?
validate selectionstart and selectionend when text is changed

Checklist:

- [x] Added unit tests (if possible)?

If the pull request fixes issue(s) list them like this:

Fixes #1954

### CODE REVIEW

**Comment by donandren** [comment]
@grokys @jkoritzinsky this PR should be ready for merge, fixes the issue completely

File: src/Avalonia.Controls/TextBox.cs [hunk 1]
```diff
@@ -212,7 +212,10 @@ public string Text
             {
                 if (!_ignoreTextChanges)
                 {
-                    CaretIndex = CoerceCaretIndex(CaretIndex, value?.Length ?? 0);
+                    var carretIndex = CaretIndex;
```
Comment (grokys) [inline src/Avalonia.Controls/TextBox.cs]: Nit: should be `caretIndex`.

File: tests/Avalonia.Controls.UnitTests/TextBoxTests.cs [hunk 2]
```diff
@@ -322,6 +321,71 @@ public void Press_Enter_Add_Custom_Newline()
             }
         }
 
+        [Fact]
+        public void SelectionEnd_Dont_Cause_Exception()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/TextBoxTests.cs]: Nit: should be `SelectionEnd_Doesnt_Cause_Exception`

File: tests/Avalonia.Controls.UnitTests/TextBoxTests.cs [hunk 3]
```diff
@@ -322,6 +321,71 @@ public void Press_Enter_Add_Custom_Newline()
...
+                target.Text = "123";
+
+                RaiseTextEvent(target, "456");
+
+                Assert.True(true);
+            }
+        }
+
+        [Fact]
+        public void SelectionStart_Dont_Cause_Exception()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/TextBoxTests.cs]: Nit: should be `SelectionStart_Doesnt_Cause_Exception`

**Review by grokys — CHANGES_REQUESTED** [review]
Looks good - just a few tiny nits.

**Review by grokys — APPROVED** [review]

**Pull request merged**